src/agent/src/device.rs

```diff
@@ -762,6 +836,45 @@ async fn vfio_device_handler(device: &Device, sandbox: &Arc<Mutex<Sandbox>>) ->
     })
 }
 
+// The VFIO AP (Adjunct Processor) device handler takes all the APQNs provided as device options
+// and awaits them. It sets the minimum AP rescan time of 5 seconds and temporarily adds that
+// amoutn to the hotplug timeout.
+#[cfg(target_arch = "s390x")]
+#[instrument]
+async fn vfio_ap_device_handler(
+    device: &Device,
+    sandbox: &Arc<Mutex<Sandbox>>,
+) -> Result<SpecUpdate> {
+    for ap_opt in device.options.iter() {
+        let opt = ap_opt.strip_prefix("ap:").ok_or_else(|| {
+            anyhow!(
+                "VFIO AP device option {} unexpectedly missed `ap:` indication",
+                ap_opt
+            )
+        })?;
+
+        let mut config = AGENT_CONFIG.write().await;
+        let hotplug_timeout = config.hotplug_timeout;
+        warn!(sl!(), "Temporarily adding {} seconds to hotplug timeout for AP devices due to rescan intervals", AP_CONFIG_TIME_SECONDS);
+        config.hotplug_timeout += time::Duration::from_secs(AP_CONFIG_TIME_SECONDS as u64 + 5);
+        // Drop write lock on config so that the uevent matcher can read from it
+        drop(config);
+        fs::write(AP_CONFIG_TIME_PATH, AP_CONFIG_TIME_SECONDS.to_string())?;
```

Comment:
Should this actually be?:

```suggestion
        fs::write(AP_CONFIG_TIME_PATH, config.hotplug_timeout.to_string())?;
```

Reply:
No :) Let me explain:
- Hotplugging, generally, can experience some delays. This is the normal `hotplug_timeout`, which defaults to 3 seconds.
- The uevent for AP devices will come in upon a rescan, which happens periodically.
- This period is configurable, down to a minimum of 5 seconds, and is configured in this line.

The `hotplug_timeout` that I want to use for AP is the "general" grace period of 3 seconds plus the up-to-5 seconds until rescan.

Reply:
I suspect this will not work quite the way you want, or at least, not reliably.  Part of the reason for 3s timeout is that there's often a ~5s timeout imposed by the layers above (e.g. CRI-O and/or k8s).  Making sure we time out first means we can give a useful error, rather than getting just a generic "container start up timed out" error.

But the point is, that increasing this timeout won't save you: if the hotplug takes too long, you'll still hit the upper levels timeout, just with a less meaningful error.  I've hit this in practice when I attempted to increase the timeout to deal with SHPC PCI hotplug (which also has a 5s delay).

Basically if yout hotplug takes this long, it's kind of fundamentally incompatible with container usage :(.  You may need to invent a different (VM only) hotplug mechanism that will complete more quickly.

Reply:
@dgibson Thank you for your review. You made some important comments, but let me look into this before anything else because it's a potential showstopper.

Reply:
I have a kernel patch that resolves this issue from a colleague specialised in this at this point, but let's conclude :arrow_down: first